pkg/controller/cdapmaster/cdapmaster_controller.go

```diff
@@ -370,12 +471,68 @@ func (b *Base) Objects(rsrc interface{}, rsrclabels map[string]string, observed,
 	}
 	var err error
 	for k, v := range configs {
-		resources, err = addConfigMapItem(master, getConfigName(master, k), labels, v, resources)
+		resources, err = addConfigMapItem(master, getResourceName(master, k), labels, v, resources)
 		if err != nil {
 			return nil, err
 		}
 	}
 
+	// User has reset the CR after failure.
+	if master.Spec.Image == master.Status.ImageToUse && master.Status.Meta.IsConditionTrue(upgradeFailed) {
+		master.Status.ClearCondition(upgradeFailed, upgradeResetMessage, upgradeResetMessage)
+	}
+
+	var item metav1.Object
+	if master.Status.UpgradeJobVersion == "" && imageToUseNeedsUpdate(master) {
+		item = k8s.GetItem(observed, &batchv1.Job{}, getUpgradeJobName(master), "default")
+	} else {
+		if !master.Status.Meta.IsConditionTrue(upgradeFailed) {
+			master.Status.ImageToUse = master.Spec.Image
+			master.Status.UserInterfaceImageToUse = master.Spec.UserInterfaceImage
+		}
+		return resources, nil
+	}
+
+	var upgradeJob *batchv1.Job
+	if item != nil {
+		upgradeJob = item.(*batchv1.Job)
+	}
+	var upgradeResources []reconciler.Object
+	resetImage := true
+	if upgradeJob == nil && !master.Status.Meta.IsConditionTrue(upgradeFailed) && imageToUseNeedsUpdate(master) {
```

Comment:
Can you add some comments about the case that each of these blocks is trying to handle?